hasher-matcher-actioner/hmalib/common/content_models.py

```diff
@@ -0,0 +1,222 @@
+    """
+    Record of action events in the action performer
+    keyed based on action_label and time of event.
+    For the time being the action_rules in the action message
+    being preformed are also stored as a list serialized json blobs
+    """
+
+    ACTION_TIME_PREFIX = "action_time#"
+    DEFAULT_PROJ_EXP = "PK, SK, ActionLabel, ActionRules"
+
+    time_performed: datetime.datetime
+    action_label: str
+    action_rules: t.List
+
+    @staticmethod
+    def get_dynamodb_action_time_key(time_performed: datetime.datetime) -> str:
+        return f"{ActionEventRecord.ACTION_TIME_PREFIX}{time_performed.isoformat()}"
+
+    def to_dynamodb_item(self) -> dict:
+        return {
+            "PK": self.get_dynamodb_content_key(self.content_id),
+            "SK": self.get_dynamodb_action_time_key(self.time_performed),
+            "ActionLabel": self.action_label,
+            "ActionRules": self.action_rules,
+        }
+
+    def to_json(self) -> t.Dict:
+        """
+        Used by '/content/action-history/<content-id>' in lambdas/api/content
+        """
+        result = asdict(self)
+        result.update(time_performed=self.time_performed.isoformat())
+        return result
+
+    @classmethod
+    def get_from_content_id(
+        cls,
+        table: Table,
+        content_id: str,
+    ) -> t.List["ActionEventRecord"]:
+
+        items = table.query(
+            KeyConditionExpression=Key("PK").eq(
+                cls.get_dynamodb_content_key(content_id)
+            )
+            & Key("SK").begins_with(cls.ACTION_TIME_PREFIX),
+            ProjectionExpression=cls.DEFAULT_PROJ_EXP,
+        ).get("Items", [])
+
+        return cls._result_item_to_action_record(items)
+
+    @classmethod
+    def _result_item_to_action_record(
+        cls,
+        items: t.List[t.Dict],
+    ) -> t.List["ActionEventRecord"]:
+        return [
+            ActionEventRecord(
+                content_id=cls.remove_content_key_prefix(
+                    item["PK"],
+                ),
+                time_performed=datetime.datetime.fromisoformat(
+                    item["SK"][len(cls.ACTION_TIME_PREFIX) :]
+                ),
+                action_label=item["ActionLabel"],
+                action_rules=item["ActionRules"],
+            )
+            for item in items
+        ]
+
+
+@dataclass
+class ContentObject(ContentObjectBase, JSONifiable):
+    """
+    Content object that stores the values related to a specific
+    content_id and content_type.
+    Note/TODO: Something like an override flag should be added to the submit API
+    as ContentRef equality check on bytes is impractical so instead
+    something like:
+     "This content_id exists, [x] overwrite? (I know what you're doing)"
+    is likely better.
+    """
+
+    CONTENT_TYPE_PREFIX = "content_type#"
+
+    content_type: str
+    content_ref: str
+    content_ref_type: str
+    additional_fields: t.Optional[t.Set]
```

Comment:
nit: prefer that set itself is not optional. Collections, if optional, can always have length 0.

```python
additional_fields: t.Set[<type>] = field(default_factory=set)
```

This prevents unnecessary things like:
```python
if content_obj.additional_fields:
  for field in content_obj.additional_fields:
    <do something>
```